common.rkt

```diff
@@ -23,50 +25,181 @@
+    (else (disunify-helper sub (cdr newsub) (cons (car newsub) acc)))))
+
+(define (disunify u v st)
+  (let* ((sub (state-sub st))
+         (diseq (state-diseq st))
+         (types (state-types st))
+         (unify-answer (unify u v st))
+         (newsub (if unify-answer (state-sub unify-answer) #f)))
+    (cond
+      ((not newsub) st)
+      ((eq? newsub sub) #f)
+      (else (state sub (extend-diseq (disunify-helper sub newsub '()) diseq) types)))))
+
+(define (diseq-simplify st)
+  (let* ((sub (state-sub st))
+         (diseq (state-diseq st))
+         (types (state-types st))
+         (st (state sub empty-diseq types)))
+    (foldl/and (lambda (=/=s st) (disunify (map car =/=s) (map cdr =/=s) st)) st diseq)))
+
+(define (foldl/and proc acc lst)
+  (if (null? lst)
+      acc
+      (let ((new-acc (proc (car lst) acc)))
+        (and new-acc (foldl/and proc new-acc (cdr lst))))))
+
+;; Type constraints
+(define (typify u type? st)
+  (let ((u (walk u (state-sub st))))
+    (if (var? u)
+        (let ((u-type (var-type-ref u (state-types st))))
+          (if u-type
+              (and (eqv? type? u-type) st)
+              (diseq-simplify (state (state-sub st)
+                                     (state-diseq st)
+                                     (extend-types u type? (state-types st))))))
+        (and (type? u) st))))
+
 
 ;; Reification
-(define (walk* tm sub)
-  (let ((tm (walk tm sub)))
+(struct Ans (term constraint) #:prefab)
+
+(define (walk* tm st)
+  (let* ((sub (state-sub st)) (tm (walk tm sub)))
     (if (pair? tm)
-        `(,(walk* (car tm) sub) .  ,(walk* (cdr tm) sub))
+        `(,(walk* (car tm) st) .  ,(walk* (cdr tm) st))
         tm)))
 (define (reified-index index)
   (string->symbol
     (string-append "_." (number->string index))))
 (define (reify tm st)
   (define index -1)
-  (walk* tm (let loop ((tm tm) (sub (state-sub st)))
-              (define t (walk tm sub))
-              (cond ((pair? t) (loop (cdr t) (loop (car t) sub)))
-                    ((var? t)  (set! index (+ 1 index))
-                               (extend-sub t (reified-index index) sub))
-                    (else      sub)))))
+  (let ((results (let loop ((tm tm) (st st))
+                   (define t (walk tm (state-sub st)))
+                   (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
+                         ((var? t)  (set! index (+ 1 index))
+                                    (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
+                         (else      st)))))
+    (let* ((walked-sub (walk* tm results))
+           (diseq (map (lambda (=/=) (cons (length =/=) =/=)) (state-diseq st)))
+           (diseq (map cdr (sort diseq (lambda (x y) (< (car x) (car y))))))
+           (st (diseq-simplify (state (state-sub st) diseq (state-types st))))
+           (diseq (walk* (map pretty-diseq (state-diseq st)) results))
+           (diseq (filter-not contains-fresh? diseq))
```

Comment:
Filter out the irrelevant diseqs first.  Then sort each diseq pair in `pretty-diseq`.